### PR TITLE
Datatype Namespacing

### PR DESCRIPTION
**Pull request by wesbland**
Move datatypes from mpid/common to the MPI layer as MPIR functionality.

### CODE REVIEW

**Comment by pavanbalaji** [comment]
@wesbland: sorry, this fell through the cracks.  do you still need me to review this?  if yes, can you rebase?

**Comment by wesbland** [comment]
It does need to get reviewed but I probably will never have time to rebase it. That would probably take quite a while and there's too many other priorities at this point.

**Comment by pavanbalaji** [comment]
OK.  I've labeled this as WIP, and removed myself from the assignee list.  Once it's ready, add me back and I'll take a look at it.

**Comment by wesbland** [comment]
Just rebased this back onto master to resolve the conflicts. I haven't even tried to recompile yet so We'll see what happens here. No doubt there have been changes in the last year that are going to cause problems.

**Comment by wesbland** [comment]
test:jenkins/ch4/all
test:jenkins/ch3/all

**Comment by raffenet** [comment]
test:jenkins/ch4/all
test:jenkins/ch3/all

**Comment by wesbland** [comment]
If I'm comparing these test results to previous results correctly, these results are consistent with previous runs so I think this PR is probably ready to be reviewed @pavanbalaji. It's another big one so you probably want to do this one after the collectives PR.

**Comment by wesbland** [comment]
I assume that we'll want to squash this whole branch down to one commit when it's ready to be merged (unless you want me to win the number of commits race forever 😄).

**Comment by wesbland** [comment]
test:jenkins/ch4/all
test:jenkins/ch3/all

**Comment by wesbland** [comment]
This has been rebased onto the latest master so it should take into account all of the `PREPEND_PREFIX` patches that went in earlier.

**Comment by raffenet** [comment]
test:jenkins/ch3/tcp
test:jenkins/ch4/ucx
test:jenkins/ch4/ofi

**Comment by raffenet** [comment]
I've done a fairly extensive rebase/squash of this PR. Should I open a new PR or is it OK to force-push over this branch? I'm a little concerned about losing the history...

**Comment by wesbland** [comment]
I always expected this to get squashed. It's up to you, but I don't care about losing the history.

**Comment by raffenet** [comment]
Ok, I'll force push over this and start the tests again.

**Comment by raffenet** [comment]
test:jenkins/ch3/tcp
test:jenkins/ch4/ucx
test:jenkins/ch4/ofi

File: CHANGES [hunk 1]
```diff
@@ -1498,7 +1498,7 @@ www.mcs.anl.gov/mpi/mpich2/mpich2_1_0_6changes.htm.
   output.  Otherwise, the output message will be base on the error message
   associated with the mpi_errno parameter.
 
-- MPID_Segment_init() now takes an additional boolean parameter that specifies
+- MPIR_Segment_init() now takes an additional boolean parameter that specifies
```
Comment (wesbland) [inline CHANGES]: Looking back, I don't know if this should be changed or not (or if we really care).

Comment (raffenet) [inline CHANGES]: Oh yeah, I meant to remove that...

Comment (wesbland) [inline CHANGES]: That's also the extent of my re-review as I'm way to lazy to go back over a 15k line PR. 😄

Comment (raffenet) [inline CHANGES]: I'll add it in a fixup after the tests complete. I also left off one commit from your original branch. Trying to dig it up now.

Comment (wesbland) [inline CHANGES]: Wow. Good luck finding that needle in the haystack.

Comment (raffenet) [inline CHANGES]: It was `Change sizep to DLOOP_Size * from int *`.

Comment (raffenet) [inline CHANGES]: And now after looking at it again, I originally thought you were doing the reverse, which is why I left it out. Now it looks like the right thing to do, so I'll look at adding it in the next revision. It'll be it's own commit and not part of the massive refactor commit.

**Comment by raffenet** [comment]
I dunno what happened to the last round of tests. Anyway here is a new version.
test:jenkins/ch3/tcp
test:jenkins/ch4/ucx
test:jenkins/ch4/ofi

**Comment by raffenet** [comment]
I think we are good to merge this. The only comment from @pavanbalaji was to make sure the pack/unpack utils could be overridden. I don't see any reason they can't here, since it's up to the device components to decide when/how to pack.

**Comment by wesbland** [comment]
Sounds good. Let me know if you need anything else on my end.

**Pull request merged**